File: PSO.kt
import org.jetbrains.kotlinx.multik.api.*
import org.jetbrains.kotlinx.multik.ndarray.data.*
import org.jetbrains.kotlinx.multik.ndarray.operations.*
import kotlin.random.Random

class PSO(val f: (MultiArray<Double, D1>) -> Double, val nPart: Int, val dim: Int, val w: Double, val c1: Double,
          val c2: Double, val mini: Double, val maxi: Double) {
    val X = mk.empty<Double, D2>(nPart, dim).map{ (maxi - mini) * Random.nextDouble() + mini }
    val V = mk.empty<Double, D2>(nPart, dim)
    var pBest = X.clone()
    var fit = (0 until nPart).map{f(X[it])}
    var pBestValue = (0 until nPart).map{f(pBest[it])}
    var gBest: Int = fit.indexOf(fit.minOrNull())
    var gBestValue = f(X[gBest])

    fun atualizaVel() {
        for (i in 0 until nPart){
            val r1 = Random.nextDouble()
            val r2 = Random.nextDouble()
            V[i] = (V[i] * w) + (c1 * r1 * (pBest[i] - X[i])) + (c2 * r2 * (X[gBest] - X[i]))
        }
    }

    fun atualizaPart() {
//         avalia condições
        for (i in 0 until nPart){
            X[i] = X[i] + V[i]
            for (j in 0 until dim){
                if (X[i,j] < mini){
                    X[i, j] = mini
                }
                else if (X[i, j] > maxi){
                    X[i, j] = maxi
                }
            }
        }
    }

//    fun fitness(f: (MultiArray<Double, D1>) -> Double) {
    fun fitness(){
        // calculando fitness
        fit = (0 until nPart).map{ f(X[it]) }
        // avaliando solucoes
        for (i in 0 until nPart) {
            if (fit[i] < pBestValue[i]){
                pBest[i] = X[i]
            }
            if (fit[i] < gBestValue){
                gBest = i
                gBestValue = fit[i]
            }
        }
        pBestValue = (0 until nPart).map{f(pBest[it])}
    }
}

File: Pso.cs
using System;
using System.Linq;

namespace Hello
{
 
    public class PsoNet
    {
        private Func<double[], double> f;
        public int numPart, dim, gBest;
        public double w, c1, c2, mini, maxi, gBestValue;
        public double[,] X, V, pBest;
        public double[] fit, pBestValue;
        private double g0;
        public double[] Slice(double[,] x, int row)
        {
            var n = x.GetLength(1);
            var res = new double[n];
            for (var i = 0; i < n; i++)
            {
                res[i] = x[row, i];
            }
            return res;
        }
        
        public PsoNet(Func<double[], double> F, int n, int d, double W, double C1, double C2, double Mini, double Maxi)
        {
            f = F;
            numPart = n;
            dim = d;
            w = W;
            c1 = C1;
            c2 = C2;
            mini = Mini;
            maxi = Maxi;
            
            g0 = 1.0e+8;
            var rand = new Random();
            X = new double[numPart,dim];
            pBest = new double[numPart,dim];
            V = new double[numPart,dim];
            fit = new double[numPart];
            pBestValue = new double[numPart];
            
            for (var i = 0; i < numPart; i++)
            {
                for (var j = 0; j < dim; j++)
                {
                    var temp = (maxi - mini) * rand.NextDouble() + mini;
                    X[i, j] = temp;
                    pBest[i, j] = temp;
                    V[i, j] = 0.0;
                }
            }
            for (var i = 0; i < numPart; i++)
            {
                var xtemp = Slice(X, i);
                g0 = f(xtemp);
                fit[i] = g0;
                pBestValue[i] = g0;
            }
            
            gBest = Array.IndexOf(this.fit, fit.Min()); 
            gBestValue = fit[gBest];
        }
        
        public void AtualizaVelocidade()
        {
            for (var i = 0; i < numPart; i++)
            {
                var rand = new Random();
                var r1 = rand.NextDouble();
                var r2 = rand.NextDouble();
                for (var j = 0; j < dim; j++)
                {
                    V[i, j] = w * V[i,j] + (c1 * r1 * (pBest[i,j] - X[i,j])) + (c2 * r2 * (X[gBest, j] - X[i,j]));
                }
            }
        }

        public void AtualizaParticulas()
        {
            for (var i = 0; i < numPart; i++)
            {
                for (var j = 0; j < dim; j++)
                {
                    X[i,j] = X[i,j] + V[i,j];    
                }
            }
            
            for (var i = 0; i < numPart; i++)
            {
                for (var j = 0; j < dim; j++)
                {
                    if (X[i, j] < mini)
                    {
                        X[i, j] = mini;
                    }
                    else if (X[i, j] > maxi)
                    {
                        X[i, j] = maxi;
                    }
                }
            }
            
            
        }
        
        public void fitness()
        {
            var fitemp = 0.0;
            double[] xtemp = new double[numPart];
            for (var i = 0; i < numPart; i++)
            {
                xtemp = Slice(X, i);
                fitemp = f(xtemp);
                fit[i] = fitemp;
            }
            for (var i = 0; i < numPart; i++)
            {
                fitemp = fit[i];
                if (fitemp < pBestValue[i])
                {
                    pBestValue[i] = fitemp;
                    for (var j = 0; j < dim; j++)
                    {
                        pBest[i,j] = X[i,j];
                    }
                }
                if (fitemp < gBestValue)
                {
                    gBest = i;
                    gBestValue = fit[i];
                }
            }
        }
        
    }
}